libmscore/changeMap.h

```diff
@@ -0,0 +1,103 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2009-2019 Werner Schweer
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#ifndef __CHANGEMAP_H__
+#define __CHANGEMAP_H__
+
+#include "fraction.h"
+
+/**
+ \file
+ Definition of class ChangeMap.
+*/
+
+namespace Ms {
+
+enum class ChangeMethod : signed char {
+      NORMAL,
+      EXPONENTIAL,
+      EASE_IN,
+      EASE_OUT,
+      EASE_IN_OUT       // and shake it all about
+      };
+
+enum class ChangeDirection : signed char {
+      INCREASING,
+      DECREASING
+      };
+
+//---------------------------------------------------------
+///   ChangeEvent
+///   item in ChangeMap
+//---------------------------------------------------------
+
+enum class ChangeEventType : char { FIX, RAMP };
+
+class ChangeEvent {
+      // Despite storing the tick as the key of the ChangeEvent in the ChangeMap, we also store it here.
+      // Since we're not going to be changing it (well, not much, anyway), this is good because it provides
+      // multiple ways to access the tick. Keeping it in sync isn't much trouble.
+      Fraction tick;
+      int value;
+      ChangeEventType type;
+      Fraction etick;
+      ChangeMethod method;
+      ChangeDirection direction;
+      int cachedStart   { -1 };
+      int cachedEnd     { -1 };
```

Comment:
Maybe these could be renamed to cachedStartValue or cachedStartVelocity to be more explicit about what it means. My guess about that this is velocity rather than tick is currently based only on the fact that its type is int and not a Fraction.

Reply:
Yes, maybe cachedStartValue would be better since this a generic map for ramps and fixes rather than one specifically for velocity.